src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2800,30 +2800,33 @@ public override ImmutableArray<Diagnostic> GetMethodBodyDiagnostics(Cancellation
         /// </summary>
         public override ImmutableArray<Diagnostic> GetDiagnostics(CancellationToken cancellationToken = default)
         {
-            return GetDiagnostics(DefaultDiagnosticsStage, true, cancellationToken);
+            return GetDiagnostics(DefaultDiagnosticsStage, true, symbolFilter: null, cancellationToken);
         }
 
-        internal ImmutableArray<Diagnostic> GetDiagnostics(CompilationStage stage, bool includeEarlierStages, CancellationToken cancellationToken)
+        internal ImmutableArray<Diagnostic> GetDiagnostics(CompilationStage stage, bool includeEarlierStages, Predicate<ISymbolInternal>? symbolFilter, CancellationToken cancellationToken)
         {
             var diagnostics = DiagnosticBag.GetInstance();
-            GetDiagnostics(stage, includeEarlierStages, diagnostics, cancellationToken);
+            GetDiagnostics(stage, includeEarlierStages, diagnostics, symbolFilter, cancellationToken);
             return diagnostics.ToReadOnlyAndFree();
         }
 
         internal override void GetDiagnostics(CompilationStage stage, bool includeEarlierStages, DiagnosticBag diagnostics, CancellationToken cancellationToken = default)
+            => GetDiagnostics(stage, includeEarlierStages, diagnostics, symbolFilter: null, cancellationToken);
+
+        internal void GetDiagnostics(CompilationStage stage, bool includeEarlierStages, DiagnosticBag diagnostics, Predicate<ISymbolInternal>? symbolFilter, CancellationToken cancellationToken)
         {
             var builder = BindingDiagnosticBag.GetInstance(withDiagnostics: true, withDependencies: false);
             Debug.Assert(builder.DiagnosticBag is { });
 
-            GetDiagnosticsWithoutFiltering(stage, includeEarlierStages, builder, cancellationToken);
+            GetDiagnosticsWithoutSeverityFiltering(stage, includeEarlierStages, builder, symbolFilter, cancellationToken);
 
             // Before returning diagnostics, we filter warnings
             // to honor the compiler options (e.g., /nowarn, /warnaserror and /warn) and the pragmas.
             FilterAndAppendDiagnostics(diagnostics, builder.DiagnosticBag, cancellationToken);
             builder.Free();
         }
 
-        private void GetDiagnosticsWithoutFiltering(CompilationStage stage, bool includeEarlierStages, BindingDiagnosticBag builder, CancellationToken cancellationToken)
+        private void GetDiagnosticsWithoutSeverityFiltering(CompilationStage stage, bool includeEarlierStages, BindingDiagnosticBag builder, Predicate<ISymbolInternal>? symbolFilter, CancellationToken cancellationToken)
```

Comment:
Renamed for clarity: we are now filtering, but not by _severity_ of the diagnostic. Wrapper methods take care of that.